packages/jest-config/src/utils.ts

```diff
@@ -69,29 +69,46 @@ export const replaceRootDirInPath = (
   );
 };
 
-// TODO: Type as returning same type as input
-const _replaceRootDirInObject = (
+const _replaceRootDirInObject = <T extends ReplaceRootDirConfigObj>(
   rootDir: Config.Path,
-  config: any,
-): {[key: string]: unknown} => {
-  if (config !== null) {
-    const newConfig: {[key: string]: unknown} = {};
-    for (const configKey in config) {
-      newConfig[configKey] =
-        configKey === 'rootDir'
-          ? config[configKey]
-          : _replaceRootDirTags(rootDir, config[configKey]);
-    }
-    return newConfig;
+  config: T,
+): T => {
+  const newConfig = {} as T;
+  for (const configKey in config) {
+    newConfig[configKey] =
+      configKey === 'rootDir'
+        ? config[configKey]
+        : _replaceRootDirTags(rootDir, config[configKey]);
   }
-  return config;
+  return newConfig;
+};
+
+type ReplaceRootDirConfigObj = {[key: string]: unknown};
+
+type ReplaceRootDirTags = {
+  <T extends ReplaceRootDirConfigObj>(rootDir: Config.Path, config: T): T;
+  <T extends RegExp>(rootDir: Config.Path, config: T): T;
+  <T extends Config.Path>(rootDir: Config.Path, config: T): T;
+  <T extends Array<ReplaceRootDirConfigObj>>(
+    rootDir: Config.Path,
+    config: T,
+  ): T;
+  <T extends Array<string>>(rootDir: Config.Path, config: T): T;
+  // Return any other kind
+  <T>(rootDir: Config.Path, config: T): T;
 };
 
-// TODO: Type as returning same type as input
-export const _replaceRootDirTags = (rootDir: Config.Path, config: any): any => {
+export const _replaceRootDirTags: ReplaceRootDirTags = (
+  rootDir: Config.Path,
+  config: any,
```

Comment:
remove the type annotations here?

Reply:
ts will still throw a `implicitly has an 'any' type`. 
Another option would be not using arrow function and type alias to overload, but convert to a function and overload it that way ?

Reply:
second sounds good to me!

Reply:
Will have to disable eslint, and with prettier it looks like:
```ts
/* eslint-disable import/export */
export function _replaceRootDirTags<T extends ReplaceRootDirConfigObj>(
  rootDir: Config.Path,
  config: T,
): T;
export function _replaceRootDirTags<T extends RegExp>(
  rootDir: Config.Path,
  config: T,
): T;
export function _replaceRootDirTags<T extends Config.Path>(
  rootDir: Config.Path,
  config: T,
): T;
export function _replaceRootDirTags<T extends Array<ReplaceRootDirConfigObj>>(
  rootDir: Config.Path,
  config: T,
): T;
export function _replaceRootDirTags<T extends Array<string>>(
  rootDir: Config.Path,
  config: T,
): T;
export function _replaceRootDirTags<T>(rootDir: Config.Path, config: T): T;
```
Still good with that option?

Reply:
does it work to define all the overloads, but just export the implementation? so only a single `_replaceRootDirTags` is exported.

or maybe an `export {_replaceRootDirTags}` below all the definitions if that doesn't work?

Reply:
They all must be exported. But your second option will work.

```ts

function _replaceRootDirTags<T extends ReplaceRootDirConfigObj>(
  rootDir: Config.Path,
  config: T,
): T;
function _replaceRootDirTags<T extends RegExp>(
  rootDir: Config.Path,
  config: T,
): T;
function _replaceRootDirTags<T extends Config.Path>(
  rootDir: Config.Path,
  config: T,
): T;
function _replaceRootDirTags<T extends Array<ReplaceRootDirConfigObj>>(
  rootDir: Config.Path,
  config: T,
): T;
function _replaceRootDirTags<T extends Array<string>>(
  rootDir: Config.Path,
  config: T,
): T;
function _replaceRootDirTags<T>(rootDir: Config.Path, config: T): T;
function _replaceRootDirTags(rootDir: Config.Path, config: any): any {
  if (config == null) {
    return config;
  }
  switch (typeof config) {
    case 'object':
      if (Array.isArray(config)) {
        /// can be string[] or {}[]
        return config.map(item => _replaceRootDirTags(rootDir, item));
      }
      if (config instanceof RegExp) {
        return config;
      }
      return _replaceRootDirInObject(rootDir, config);
    case 'string':
      return replaceRootDirInPath(rootDir, config);
  }
  return config;
}

export {_replaceRootDirTags};
```

Reply:
another option without overloads but having to cast:
```ts
type ReplaceRootDirConfigObj = {[key: string]: unknown};
type ReplaceRootDirConfigValues =
  | ReplaceRootDirConfigObj
  | RegExp
  | Config.Path
  | Array<ReplaceRootDirConfigObj>
  | Array<string>;

export const _replaceRootDirTags = <T extends ReplaceRootDirConfigValues>(
  rootDir: Config.Path,
  config: T,
): T => {
  if (config == null) {
    return config;
  }
  switch (typeof config) {
    case 'object':
      if (Array.isArray(config)) {
        /// can be string[] or {}[]
        return config.map(item => _replaceRootDirTags(rootDir, item)) as T;
      }
      if (config instanceof RegExp) {
        return config;
      }

      return _replaceRootDirInObject(
        rootDir,
        config as ReplaceRootDirConfigObj,
      ) as T;
    case 'string':
      return replaceRootDirInPath(rootDir, config) as T;
  }
  return config;
};
```

Reply:
That seems cleaner, no?

Reply:
Yeah def cleaner, just didnt know your tolerance when it comes to casting. i'll go with it